### ISSUE TITLE
tf.function retracing for multigpu(mirrored_strategy) operations(tf.keras.models)

### ISSUE THREAD

**Issue by wiseosho** [tensorflow/tensorflow#34609]
**System information**
- Have I [written]() custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): ('CentOS Linux', '7.6.1810', 'Core')
- os platform: Linux-3.10.0-862.14.4.el7.x86_64-x86_64-with-centos-7.6.1810-Core
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: No
- TensorFlow installed from (source or binary): No
- TensorFlow version (use command below): unknown 2.0.0
- Python version: 3.7.2
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:  cudatoolkit-10.0.130 / cudnn-7.6.4 
- GPU model and memory: TITAN Xp   (12G) * 8

**Describe the current behavior**

Currently I have 8 gpus in develop environment(titanxp's), and converting single gpu training model using tf.distributed.mirror_strategy & tf.function gives slow training(gpu uses) and print WARNINGS below.
Since I confirmed model trains well in single gpu environment, I Wrote simple test code similar to the structure of my actual code of my model, which shows the same WARNINGS.

`W1126 17:46:13.538742 139942541895424 def_function.py:474] 5 out of the last 5 calls to <function inference at 0x7f4a76ab12f0> triggered tf.function retracing. Tracing is expensive and the excessive number of tracings is likely due to passing python objects instead of tensors. Also, tf.function has experimental_relax_shapes=True option that relaxesargument shapes that can avoid unnecessary retracing. Please refer to https://www.tensorflow.org/beta/tutorials/eager/tf_function#python_or_tensor_args and https://www.tensorflow.org/api_docs/python/tf/function for more details.
W1126 17:46:13.584210 139942533502720 def_function.py:474] 6 out of the last 6 calls to <function inference at 0x7f4a76ab12f0> triggered tf.function retracing. Tracing is expensive and the excessive number of tracings is likely due to passing python objects instead of tensors. Also, tf.function has experimental_relax_shapes=True option that relaxesargument shapes that can avoid unnecessary retracing. Please refer to https://www.tensorflow.org/beta/tutorials/eager/tf_function#python_or_tensor_args and https://www.tensorflow.org/api_docs/python/tf/function for more details.
W1126 17:46:13.634693 139941979879168 def_function.py:474] 7 out of the last 7 calls to <function inference at 0x7f4a76ab12f0> triggered tf.function retracing. Tracing is expensive and the excessive number of tracings is likely due to passing python objects instead of tensors. Also, tf.function has experimental_relax_shapes=True option that relaxesargument shapes that can avoid unnecessary retracing. Please refer to https://www.tensorflow.org/beta/tutorials/eager/tf_function#python_or_tensor_args and https://www.tensorflow.org/api_docs/python/tf/function for more details.
W1126 17:46:13.689692 139941971486464 def_function.py:474] 8 out of the last 8 calls to <function inference at 0x7f4a76ab12f0> triggered tf.function retracing. Tracing is expensive and the excessive number of tracings is likely due to passing python objects instead of tensors. Also, tf.function has experimental_relax_shapes=True option that relaxesargument shapes that can avoid unnecessary retracing. Please refer to https://www.tensorflow.org/beta/tutorials/eager/tf_function#python_or_tensor_args and https://www.tensorflow.org/api_docs/python/tf/function for more details.`

I wonder if this retracing could occur from restricted network policy or outdated kernels. Since this workstation is originally intended for the service purpose so it is under strong monitoring for network packets(ports). I found mirrored_strategy open several network ports randomly at startup.

**Describe the expected behavior**
No retracing warnings with fully dedicated GPU usages.

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.
```python
import tensorflow as tf
import os, sys
from tensorflow.keras import layers
from tensorflow.python.keras import backend, layers, models, utils
from os.path import join, basename, dirname
from os import listdir
tf.__version__
tf.executing_eagerly()

mirrored_strategy = tf.distribute.MirroredStrategy()#devices=["/gpu:0", "/gpu:1"])#devices=["/gpu:0", "/gpu:1"])#,
                                                  #cross_device_ops=tf.distribute.HierarchicalCopyAllReduce())
    
def create_model():
  model = tf.keras.Sequential([
      tf.keras.layers.Conv2D(32, 3, activation='relu'),
      tf.keras.layers.MaxPooling2D(),
      tf.keras.layers.Conv2D(64, 3, activation='relu'),
      tf.keras.layers.MaxPooling2D(),
      tf.keras.layers.Flatten(),
      tf.keras.layers.Dense(64, activation='relu'),
      tf.keras.layers.Dense(10, activation='softmax')
    ])

  return model


with mirrored_strategy.scope():
    
    inp = layers.Input([240,320,6])
    
    
    test = create_model()
    

@tf.function
def inference(input):
    return test(input)
@tf.function
def inf():
    with mirrored_strategy.scope():
            mirrored_strategy.experimental_run_v2(inference, args=(tf.random.normal(shape=[32,240,320,3]),)) 

inf()

```
**Other info / logs**
Although I remove the loss calculation and optimizations, I checked this retracing occurs irrelevant to the loss or optimization steps.

**Comment by wiseosho** [comment]
I checked the code on gcp compute instance with K80*8 gpu environment, and it works fine.
(Same anaconda environment with tensorflow-gpu 2.0.) 
Now I guess this issue is caused by my development environment, but I still wonder why my environment cause retracing problem for multigpu training.

**Comment by guptapriya** [comment]
I am not able to repro this either. Is it possible that you have somewhat different versions of TF? I tested with TF @ master.

**Comment by wiseosho** [comment]
No, I am using tensorflow provided by anaconda.

**Comment by wiseosho** [comment]
I would appreciate a effective way for tracing the problem.
It seems the tf.function decorator spawn another process so actually impossible to trace directly from my code.

**Comment by guptapriya** [comment]
tf.function does not spawn another process - why do you say that?

**Comment by wiseosho** [comment]
Currently I am using another deeplearning framework and it runs smoothly without issue. I cannot stop training to reproduce the results or recheck your comments since training occupies all gpus available.
You could close the issue or I will update anytime one I recheck the debugging comments.
Thank you for your kind support.
